File: recipe-manager.backend/src/RecipeManager.Data/Entities/Recipe.cs
namespace RecipeManager.Data.Entities;

public class Recipe
{
    public int Id { get; init; }
    public string Title { get; init; } = string.Empty;
    public int Servings { get; init; }
    public int CaloriesPerServing { get; init; }
    public double ProteinG { get; init; }
    public double CarbsG { get; init; }
    public double FatG { get; init; }
    public string? PrimaryProtein { get; init; }
    public DateTime CreatedAt { get; init; }
    public DateTime UpdatedAt { get; init; }

    public IReadOnlyList<Tag> Tags { get; init; } = [];
    public IReadOnlyList<RecipeStep> Steps { get; init; } = [];
}

File: recipe-manager.frontend/src/types/recipe.ts
export interface Tag {
  id: number;
  name: string;
}

export interface CulinaryAction {
  id: number;
  name: string;
}

export interface RecipeStepIngredient {
  id: number;
  ingredientName: string;
  quantity: number;
  unit: string;
}

export interface RecipeStep {
  id: number;
  orderIndex: number;
  actions: string[];
  ingredients: RecipeStepIngredient[];
  durationMinutes: number | null;
  notes: string | null;
}

export interface Recipe {
  id: number;
  title: string;
  servings: number;
  caloriesPerServing: number;
  proteinG: number;
  carbsG: number;
  fatG: number;
  primaryProtein: string | null;
  tags: Tag[];
  steps: RecipeStep[];
}

export interface CreateStepIngredientRequest {
  ingredientName: string;
  quantity: number;
  unit: string;
}

export interface CreateStepRequest {
  orderIndex: number;
  actions: string[];
  ingredients: CreateStepIngredientRequest[];
  durationMinutes: number | null;
  notes: string | null;
}

export interface CreateRecipeRequest {
  title: string;
  servings: number;
  caloriesPerServing: number;
  proteinG: number;
  carbsG: number;
  fatG: number;
  primaryProtein: string | null;
  tags: string[];
  steps: CreateStepRequest[];
}

export type UpdateRecipeRequest = CreateRecipeRequest;
